04_basic_struct/bstructmodule.c

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: GPL-3.0
+/**/
+
+#include <linux/init.h>	// Macros used to mark up functions  __init __exit
+#include <linux/module.h> // Core header for loading LKMs into the kernel
+#include <linux/kobject.h>
+#include <linux/err.h>
+#include <linux/string.h>
+#include <linux/sysfs.h>
+#include <linux/list.h>
+#include <linux/slab.h>
+
+struct list_node {
+	char *node;
+	struct list_head head_list;
+};
+
+LIST_HEAD(linkedlist);
+
+static ssize_t hello_show(struct kobject *kobj, struct kobj_attribute *attr,
+	char *buf)
+{
+	struct list_node *cur_node = NULL;
+	int offset = 0;
+	int res = 0;
+
+	list_for_each_entry(cur_node, &linkedlist, head_list) {
+		pr_info("bstructmodule: show nodes: %s\n", cur_node->node);
+		res = scnprintf(buf+offset, PAGE_SIZE, "%s\n", cur_node->node);
+		offset += res;
+	}
+	return offset;
+}
+
+static ssize_t hello_store(struct kobject *kobj, struct kobj_attribute *attr,
+	const char *buf, size_t count)
+{
+	struct list_node *new_node =
+		kmalloc(sizeof(struct list_node), GFP_KERNEL);
+
+	if (!new_node)
+		return -ENOMEM;
+	new_node->node = kmalloc_array(count, sizeof(char), GFP_KERNEL);
+	strcpy(new_node->node, buf);
+	list_add(&new_node->head_list, &linkedlist);
+	pr_info("bstructmodule: store nodes: Added node = %s\n", buf);
+	return count;
+}
+
+static struct kobj_attribute list_attribute =
+	__ATTR(list, 0664, hello_show, hello_store);
+
+static struct attribute *attrs[] = {
+	&list_attribute.attr,
+	NULL,
+};
+
+static struct attribute_group attr_group = {
+	.attrs = attrs,
+};
+
+static struct kobject *hello_kobj;
+
+static int __init hello_init(void)
+{
+	int res = 0;
+
+	pr_debug("bstructmodule: loading started\n");
+	hello_kobj = kobject_create_and_add("hello", kernel_kobj);
+	if (!hello_kobj)
+		return -ENOMEM;
+	/* Create the files associated with this kobject */
+	res = sysfs_create_group(hello_kobj, &attr_group);
+	if (res)
+		kobject_put(hello_kobj);
+	return res;
+}
+
+static void __exit hello_exit(void)
+{
+	struct list_node *cur_node = NULL;
+	struct list_node *tmp_node = NULL;
+
+	list_for_each_entry_safe(cur_node, tmp_node, &linkedlist, head_list) {
+		list_del(&cur_node->head_list);
+		kfree(cur_node->node);
+		kfree(cur_node);
+	}
+	kobject_put(hello_kobj);
+	pr_info("bstructmodule: module exited\n");
```

Comment:
Read about [pr_fmt](https://www.kernel.org/doc/html/latest/core-api/printk-basics.html) 
Try to add
```c
#define pr_fmt(fmt) "%s: " fmt ,  KBUILD_MODNAME
```
before all `#include`-s and remove `"bstructmodule: "` substring from every format string in this module
```c
pr_info("module exited\n");
```
Think about how this C magic works.

Reply:
Thank you for review. I've already fix code.